File: src/codes/hardcoded-variability/py/rfc_strategy.py
from abc import ABC, abstractmethod

class NotificationStrategy(ABC):
    @abstractmethod
    def send(self, user: str, message: str):
        pass

class EmailStrategy(NotificationStrategy):
    def send(self, user: str, message: str):
        print("件名: お知らせ")
        print(f"本文: {message}")
        print(f"宛先: {user}@example.com")

class SlackStrategy(NotificationStrategy):
    def send(self, user: str, message: str):
        print(f"Slack宛: {user}")
        print(f"内容: {message}")

class NotificationService:
    def __init__(self, strategy: NotificationStrategy):
        self.strategy = strategy

    def notify(self, user: str, message: str):
        self.strategy.send(user, message)

# 利用例
NotificationService(EmailStrategy()).notify("hiroshi", "メールメッセージ")
NotificationService(SlackStrategy()).notify("hiroshi", "Slackメッセージ")

File: src/codes/hardcoded-variability/ts/rfc_strategy.ts
// Strategy インターフェース
interface NotificationStrategy {
  send(user: string, message: string): void;
}

// Strategy 実装（メール）
class EmailStrategy implements NotificationStrategy {
  send(user: string, message: string): void {
    console.log("件名: お知らせ");
    console.log(`本文: ${message}`);
    console.log(`宛先: ${user}@example.com`);
  }
}

// Strategy 実装（Slack）
class SlackStrategy implements NotificationStrategy {
  send(user: string, message: string): void {
    console.log(`Slack宛: ${user}`);
    console.log(`内容: ${message}`);
  }
}

// コンテキスト
class NotificationService {
  constructor(private strategy: NotificationStrategy) {}

  notify(user: string, message: string) {
    this.strategy.send(user, message);
  }
}

// 利用例
new NotificationService(new EmailStrategy()).notify(
  "hiroshi",
  "メールメッセージ"
);
new NotificationService(new SlackStrategy()).notify(
  "hiroshi",
  "Slackメッセージ"
);
